### ISSUE TITLE
Add deprecation or change notices for API consumers

### ISSUE THREAD

**Issue by ijdickinson** [epimorphics/elda#165]
From time-to-time, we necessarily have to update an Elda API due to changes in data-model or other business changes. Since Elda is often used to provide public APIs that users are not required to register to use, we typically don't have any direct means to contact API users to tell them that a deprecation or breaking change has occurred or will be occuring.

It would be helpful to add to the Elda configuration some means of conveying changes to API users. This could manifest in several ways:

* as a custom property in the metadata section of the JSON, Turtle or XML output (parenthetically, those are in priority order for my personal use cases)

* as part of the `Page` object, with a default rendering in the Velocity and XSLT rendered pages

* as a section on the API config page

There may be other options.

I would expect the inline change notification to be fairly brief; longer change notices could be provided as links if the change notification was in HTML or Markdown.

**Comment by ehedgehog** [comment]
As an experiment I have a version of Elda that defines a property elda:deprecated that can
be applied to an API or any of its endpoints and which is (recursively) copied into the page 
metadata. This makes it available in the metadata sections of the JSON, XML, or Turtle
renderings.

The XSLT/HTML and Velocity/HTML renderings wouldl require some support code and
UI decisions about where deprecation notices should appear. Perhaps as another 
item on the page header which understands some specific properties of the deprecation
object?

**Comment by ijdickinson** [comment]
Thanks Chris. Do you have any examples, or tests, or documentation patches that would show what you've got so far?

**Comment by bwmcbride** [comment]
Please post examples of configuration and XML and JSON output.

elda:deprecated suggests that the api or endpoint is already deprecated, whereas the requirement is also to give notice of upcoming deprecations.  What about a more general elda:notice property?  Maybe also use the owl:deprecated property to indicated actual deprecation.

**Comment by ehedgehog** [comment]
We can call it whatever we like, and if we want more than 1 property
producing notices, we can do that too. How about elda:notice as
the property, with elda:Notice as its range type?

**Comment by ehedgehog** [comment]
Example. Suppose we have an endpoint with a deprecation in it (keeping
the old name for now):

<pre>
ello:games a api:ListEndpoint 
    ; rdfs:label "Games" 
    ; api:uriTemplate "/games" 
    ; api:selector [api:filter "type=BoardGame"; api:sort "label"]
    ; api:defaultViewer api:labelledDescribeViewer
    ; elda:enableCounting false
    ; elda:deprecated &lt;eh:/this.is.deprecated-B>
    .
</pre>
    
&lt;eh:/this.is.deprecated-B> a elda:Notice
    ; rdfs:label "this is deprecated (policy B)"
    ; elda:startDate "2016-06-01"^^xsd:date
    .

**Comment by ehedgehog** [comment]
(I made up elda:startDate because I didn't want to go on a vocabulary
hunt; I'm sure there's a property we can exploit in some well-known
vocabulary. The experimental code doesn't care what the properties are
when it copies them into the metadata model, but obviosuly the html 
renderers will need an agreed set of properties to display.)

**Comment by bwmcbride** [comment]
So if we were using elda:notice rather than elda:deprecated, and keeping the range to a simple literal,  this might look like:

```
ello:games a api:ListEndpoint 
    ; rdfs:label "Games" 
    ; api:uriTemplate "/games" 
    ; api:selector [api:filter "type=BoardGame"; api:sort "label"]
    ; api:defaultViewer api:labelledDescribeViewer
    ; elda:enableCounting false
    ; elda:notice 
         """This API is now deprecated.  
               It has been replaced by the FOO api - see http://example.org/foo-api.
        """
     ; owl:deprecated true
    .
```

Trying out the idea we don't need to treat the notice as a THING, which keeps it simpler.

What does the output look like?

**Comment by ehedgehog** [comment]
The output looks like an error because the current code assumes that
the object of a deprecated (or notice) property is a resource.

If it allowed that, then we have the appropriate rendering of
the statement `thepage elda:notice "the string".

**Comment by bwmcbride** [comment]
Chris please could you post an example of what the output will look like in both XML and JSON.  Doesn't matter whether it was produced by the current code.  I think it would be sensible to look at the intended output so we can be sure we are comfortable with it.

**Comment by ehedgehog** [comment]
There's a JSON and XML (and Turtle) example above.

Note that in all three cases all that's happening is the format's usual
rendering of rdf statements.

**Comment by ijdickinson** [comment]
Hi Chris,
I agree with Brian that seeing a more complete example would be useful. In particular, a complete JSON rendering for an item endpoint with a notice attached would be helpful (I say JSON, because I think that most of the people outside of our office who we want to reach will be using the JSON output)

**Comment by ehedgehog** [comment]
(Answering this question has exposed a bug; probing ...)

**Comment by ehedgehog** [comment]
An endpoint with a notice:

<pre>
spec:school a api:ItemEndpoint
  	; rdfs:label "School"
  	; api:uriTemplate "/doc/school/{number}"
  	; api:itemTemplate "http://education.data.gov.uk/id/school/{number}"
  	; api:defaultViewer spec:viewerMedium
  	; api:viewer
    	spec:viewerProvision
    	, spec:viewerLocation
    ; elda:notice <lt;eh:/this.is.deprecated-B>
    .

<lt;eh:/this.is.deprecated-B> a elda:Notice
    ; rdfs:label "this is deprecated (policy again:B)"
    ; elda:startDate "2016-06-01"^^xsd:date
    .
</pre>

The query
<pre>
http://localhost:8080/standalone/mini/doc/school/100866.json
</pre>

Produces the output:

<pre>
{
    "format": "linked-data-api",
    "version": "0.2",
    "result": {
        "_about": "http://localhost:8080/standalone/mini/doc/school/100866.json",
        "definition": "http://localhost:8080/standalone/mini/meta/doc/school/_number.json",
        "extendedMetadataVersion": "http://localhost:8080/standalone/mini/doc/school/100866.json?_metadata=all",
        "notice": {
            "_about": "eh:/this.is.deprecated-B",
            "label": "this is deprecated (policy again:B)",
            "startDate": "Wed, 01 Jun 2016",
            "type": [
                "http://www.epimorphics.com/vocabularies/lda#Notice"
            ]
        },
        "primaryTopic": {
            "_about": "http://education.data.gov.uk/id/school/100866",
            "PFI": false,
            "address": {
                "_about": "http://api.talis.com/stores/govuk-education/items/1259149807712#self",
                "address1": "127 Herne Hill",
                "address2": "Herne Hill",
                "postcode": "SE24 9LY",
                "town": "London"
            },
            "altEmail": "vivien.tabone@hernehillschool.co.uk",
            "establishmentNumber": 6375,
            "gender": {
                "_about": "http://education.data.gov.uk/def/school/Gender_Mixed",
                "label": "Mixed"
            },
            "isPrimaryTopicOf": "http://localhost:8080/standalone/mini/doc/school/100866.json",
            "label": "Herne Hill School",
            "notes": "ISI 6 yr review date Jun '11",
            "openDate": "Fri, 17 Feb 1989",
            "prefEmail": "enquiries@hernehillschool.co.uk",
            "schoolCapacity": 275,
            "typeOfEstablishment": {
                "_about": "http://education.data.gov.uk/def/school/TypeOfEstablishment_TERM_Other_Independent_School",
                "label": "Other Independent School"
            },
            "uniqueReferenceNumber": 100866
        },
        "type": [
            "http://purl.org/linked-data/api/vocab#ItemEndpoint",
            "http://purl.org/linked-data/api/vocab#Page"
        ]
    }

}
</pre>

**Comment by ijdickinson** [comment]
Thanks Chris. That looks pretty good to me.  I guess the implication is that there's only a single notice at a time, which would be updated to include new information as it becomes available (and then given a new time-stamp). I think I prefer the simplicity of a single notice to the relative complexity of making it a list.

**Comment by ehedgehog** [comment]
There's ony one notice in the example, so there's only one notice
in the metadata. If there are multiple notices on a page, then
the value of elda:notice will be a list, which is the default
behaviour for properties-to-json in Elda.

It should be easy to make it an array even when there's 
only one value.

**Comment by ijdickinson** [comment]
As a developer, I find the it-might-be-a-list-or-it-might-not-be to be really inconvenient.  As I say though, I'm in favour of keeping it simple so that it is just a single value. Apart from anything else, updating the UI to show *the* notice is one thing, making it show a list of the N notices, which may or may not be ordered by date, is something quite different.

**Comment by ehedgehog** [comment]
Well. If there's only one notice on the endpoint, then there's only
one value. Should we specifically disallow multiple notices on
an endpoint/API? I grant that handling multiple notices in the
UI may be tricky.

Apropos, how shall we present a notice in the HTML UIs?

**Comment by ehedgehog** [comment]
For Elda 1.3.18 we have introduced the config property `elda:notice`.
It is copied with its properties and values from the LDA config API and
endpoint to the page metadata where it can be extracted by an Elda
client.

No UI is current defined for this feature; see issue #166.

**Comment by ehedgehog** [comment]
elda:notice is in the released 1.3.18.
